### PR TITLE
feat(secretsmanager): import secrets by name

### LINKED ISSUE #7444: SecretsManager: import by name, add secretId property to ISecret
The Secrets Manager import feature is currently quite annoying to import existing secrets in an account-agnostic way.

It would be perfectly fine to import and access them by name, but the API doesn't currently support it.

Needs an accessor to hold the non-ARN attribute, and needs to be calculcated from an ARN if not available, and we need to grant permissions on `<id>-*`.

---

This is a :rocket: Feature Request

### LINKED ISSUE #7949: Get the secretmanager secret Arn
I am looking to get the arn of a secret so I can allow infrastructure I am deploying to have the permissions to get that secret e.g.:

const secret = secretsmanager.Secret.fromSecretArn(stack, secretName, secretArn);
 secret.grantRead(CdkBuild);

In this case to allow a codebuild build to be able to read a secret without me having to pass it in a way which appears in the logs (e.g. echo to file or... etc)

I can get the arn by using the AWS SDK listSecrets call and then iterating until the secretName matches but that is an async call and the value is returned after the cdk has passed the point where it needs the value.

I can get the secret value itself from within cdk e.g. for a github token:

ghtokenSecret = cdk.SecretValue.secretsManager(secretName, {
    jsonField: ghtoken,
 });

But can see no method/way to get the secret arn such that I can pass the value to my codebuild infrastructure.

### Use Case

Given multiple accounts for multiple environments (dev, sandbox, integration etc) allowing the use of just the SecretName allows me to use the cdk script effectively unchanged vs if using the arn as I have to create the secret for that account (using the same secret name) and then find it's arn and then insert that arn into some config (e.g. cdk.json) for each branch etc. which is potentially a source of failure/error.

This allows the secret to stay secret as opposed to echo'ing the value which appears in the cloudformation logs etc.

### Proposed Solution

Basically I would like something like:

const secret = secretsmanager.Secret.fromSecretName(stack, secretName);

and then secret has a .arn property

Though if I could get the secret using the above I might not need the arn, however it would fit the pattern of other arn based infrastructure.

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### LINKED ISSUE #7994: Support Secret with partial Arn
When a container in a `TaskDefinition` uses a secret, it automatically adds to the taskDefinition generated `ExecutionRole` the IAM [permission to read the secret value](https://github.com/aws/aws-cdk/blob/ca90f5b8650815641c761eee9fa7d565ce0bc20c/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L394).

If the secret is imported from an existing resource, by using `secretsmanager.Secret.fromSecretAttributes` or `secretsmanager.Secret.fromSecretArn` the Arn can be an [partial Arn](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html#API_GetSecretValue_RequestSyntax), without the `-` and 6 random chars AWS appends to the friendly name.

The problem is that [Secret.grantRead](https://github.com/aws/aws-cdk/blob/ca90f5b8650815641c761eee9fa7d565ce0bc20c/packages/%40aws-cdk/aws-secretsmanager/lib/secret.ts#L132) gives permission to the exact Arn only, so the task fails to execute.

### Use Case

It's common to use partial Arn when using secrets, because they represent friendly names.
### Proposed Solution

I tried to manually add a more relaxed policy to the taskDefinition execution role, but it did not work.

```
taskDefinition.addToExecutionRolePolicy(new iam.PolicyStatement({
  actions: ["secretsmanager:GetSecretValue"],
  resources: [secretArn + '*'],
}));
```

I don't know why it did not work, or if this is the best approach.

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by njlynch**
Adds the ability to import secrets by name, including without the SecretsManager
assigned suffix. As long as a secret with the same name has been created in each
region with the same name, this allows for the same `fromSecretName` usage in
stacks across regions.

Oddly enough, most CloudFormation templates that take references to secrets
accept either the full-form ARN, including the suffix or just the base secret
name (not in ARN format). The one place where a full ARN format is needed is in
IAM policy statements, where the wildcard is necessary to account for the
suffix.

Tested this manually against an existing secret with a CodeBuild project; per
the CloudFormation docs, this should work equally well with other
SecretsManager-integrated services.

fixes #7444
fixes #7949
fixes #7994

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:243]: Why not a getter?

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:271]: `+ '-*'`? To make sure the only part we wildcard is the generated suffix?

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:606]: This will include the generated suffix. Is that on purpose?

**Review by rix0rrr — APPROVED** [review]
Approved provided small reconsiderations

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:243]: Sure, can do.

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:271]: I considered that, and even coded it up that way initially, then changed my mind. This impacts your other comment as well. Happy for a second opinion here.

If the user provides only the "friendly" version of the secret name (e.g., `My-Special-Secret`), then yes, we should prepend with `'-*'` to do the right thing. However, if a user provides the whole name (e.g., `My-Special-Secret-123bdc`), then adding `-*` will break the permissions; this is fairly subtle breakage, so I could see it being hard to diagnose. The gotcha is that we can't tell if the user's provided the full secret name or just the "friendly" name. 

Also, using `-*` doesn't guarantee we'll only wildcard the suffix. If I have a secret `My-Special-Secret` that I want access to, I can create another secret `My-Special`, and `grantRead` on that will give me access to `My-Special-Secret` too. This is admittedly an edge case with secret names that include `-`, but it's possible.

So in the end I weighed the benefits (slightly better, but far from foolproof security) and risks (really subtle breakage if the user supplies the "full" secret name) and opted for this. Totally willing to have my mind changed here though.

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:606]: I _could_ strip out everything after the last `-` from the resourceName; I was a bit worried about people providing "incomplete" ARNs (e.g., ARN format but without the suffix) and that breaking things. But I could always enforce that the arn must have a resourceName and that resourceName must contain at least one `-`.

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (eladb) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:138]: Why is this useful if `secretArn` is required?

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (eladb) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:262]: This should be an ARN, no?

File: packages/@aws-cdk/aws-docdb/package.json
Comment (eladb) [inline packages/@aws-cdk/aws-docdb/package.json:96]: Why is this needed?

We desperately need to require "motivation" in the exclude list!

Comment (njlynch) [reply packages/@aws-cdk/aws-docdb/package.json:96]: Because secretName isn't an attribute.

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:262]: No. We don't have an ARN here, and intentionally do **not** want to generate something that looks like one. Valid use cases for references to a secret from other services is either the full complete arn (e.g., `arn:aws:secretsmanager....:MySecret-a3g3cf`) or the shortened secret name (e.g., `MySecret`). Creating an "incomplete" ARN (e.g., `arn:aws:secretsmanager....:MySecret`) is invalid and cannot be used anywhere. The one exception is for IAM policies, where we use the `arnForPolicies` to generate an incomplete ARN with a wildcard.

File: packages/@aws-cdk/aws-secretsmanager/lib/secret.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:138]: That's a fair comment; I think it's primarily a hold-over from when `fromSecretName` used `fromSecretAttributes`, which it no longer does. I kept it for completeness, but it could just as easily be removed.

Comment (njlynch) [reply packages/@aws-cdk/aws-secretsmanager/lib/secret.ts:138]: https://github.com/aws/aws-cdk/pull/10410

**Comment by rrrix** [comment]
Uhg. This totally broke my production stack.

I'm importing a secret using from another stack by ARN, like this:

```typescript
    const dbSecret = secretsmanager.Secret.fromSecretArn(this, 'dbSecret', props.dbSecretArn);
```

I get this error:
```
/Users/me/project/cdk/node_modules/@aws-cdk/aws-secretsmanager/lib/secret.ts:605
  throw new Error('invalid ARN format; no secret name provided');
        ^
Error: invalid ARN format; no secret name provided
    at parseSecretName (/Users/me/project/cdk/node_modules/@aws-cdk/aws-secretsmanager/lib/secret.ts:605:9)
    at new Import (/Users/me/project/cdk/node_modules/@aws-cdk/aws-secretsmanager/lib/secret.ts:282:36)
    at Function.fromSecretAttributes (/Users/me/project/cdk/node_modules/@aws-cdk/aws-secretsmanager/lib/secret.ts:286:12)
    at Function.fromSecretArn (/Users/me/project/cdk/node_modules/@aws-cdk/aws-secretsmanager/lib/secret.ts:245:19)
    at new APIService (/Users/me/project/cdk/src/services/api.ts:77:44)
    at main (/Users/me/project/cdk/src/index.ts:190:22)
```


@njlynch What's supposed to happen if the `secretArn` passed here is a Token (e.g. doesn't exist yet?)

```typescript
/** Returns the secret name if defined, otherwise attempts to parse it from the ARN. */
export function parseSecretName(construct: IConstruct, secretArn: string, secretName?: string) {
  if (secretName) { return secretName; }
  const resourceName = Stack.of(construct).parseArn(secretArn).resourceName;
  if (resourceName) {
    // Secret resource names are in the format `${secretName}-${SecretsManager suffix}`
    const secretNameFromArn = resourceName.substr(0, resourceName.lastIndexOf('-'));
    if (secretNameFromArn) { return secretNameFromArn; }
  }
  throw new Error('invalid ARN format; no secret name provided');
}
```

```shell
$ cdk --version
1.64.0 (build 9510201)
```

**Comment by Cloudrage** [comment]
Same here since update in 1.64.0 :
```
invalid ARN format; no secret name provided
Subprocess exited with error 1
```

**Comment by snorberhuis** [comment]
This also breaks our production stack.

**Comment by Cloudrage** [comment]
To prevent that you have to set a fix version of CDK.
When a new version came available, update it but test on your dev/stg environments before updating your prd.

I'm using pipelines in most cases and only with a valid/fixed CDK version; in that case you don't have any surprise.
With the actual SM pb, I've seen it in my DEV environment and came back to the previous version (this version did not contain any expected fixed on my side). Or just the SM package version.

**Pull request merged**
